waiter/src/waiter/state.clj

```diff
@@ -1180,32 +1180,34 @@
 
 (defn start-router-state-maintainer
   "Start the instance state maintainer.
-   Maintains the state of the router as well as the state of marathon
-   and the existence of other routers. Acts as the central access point
-   for modifying this data for the router."
+   Maintains the state of the router as well as the state of marathon and the existence of other routers.
+   Acts as the central access point for modifying this data for the router.
+   Exposes the state of the router via a `query-state-fn` no-args function that is returned."
   [scheduler-state-chan router-chan router-id exit-chan service-id->service-description-fn deployment-error-config]
-  (let [state-chan (async/chan)
+  (let [state-atom (atom {:all-available-service-ids #{}
```

Comment:
What do you think about keeping the initialization inside the loop, and keeping the loop such that it doesn't ever read the atom?  That seems cleaner to me.  Then just `(reset! state-atom state)` (`state` is the loop var)

Reply:
Initializing the atom as we do current avoids us having to initialize the atom to a `nil` value and worry about invocations that retrieve the state even before the go-block has had a chance to run and reset the value in the atom.

Reply:
That seems like a problem with the `atom` approach as written, no?  Previously, did we wait for first available value?

Reply:
Yes, it is a side-effect of going down the atom approach where we want the atom to reflect the daemon state (including the initial state). Previously we would read from the query channel and the go-block would respond only after it got started.

Reply:
Btw, would you be happier if I created a variable `initial-state` in the `let` block wrapping the loop and used it in the loop initialization? Then the atom could be reset to this initial state outside the go-block. An example is here: https://github.com/twosigma/waiter/pull/408/commits/41d484af4e13ca6e006a7c08e847d4533995dd17#diff-4afda66f399219b670b4e625088d4eecR1187